File: src/main/java/UdfFactory.java
import org.apache.spark.sql.expressions.UserDefinedFunction;
import org.apache.spark.sql.types.DataType;
import org.apache.spark.sql.types.DataTypes;

import static org.apache.spark.sql.functions.udf;

import java.util.Random;

public class UdfFactory {
    public static UserDefinedFunction createMapFunction(String functionName, String[] inputColumnNames, String outputColumnType){
        DataType outputDataType = DataType.fromDDL(outputColumnType);
        switch(inputColumnNames.length) {
            case 0:
                return udf(
                        () -> {
                            return FunctionWrapper.get().apply()[0];
                        },
                        outputDataType);
            case 1:
                return udf(
                        (a1) -> {
                            return FunctionWrapper.get().apply(a1)[0];
                        },
                        outputDataType);
            case 2:
                return udf(
                        (a1, a2) -> {
                            return FunctionWrapper.get().apply(a1, a2)[0];
                        },
                        outputDataType);
            case 3:
                return udf(
                        (a1, a2, a3) -> {
                            return FunctionWrapper.get().apply(a1, a2, a3)[0];
                        },
                        outputDataType);
            case 4:
                return udf(
                        (a1, a2, a3, a4) -> {
                            return FunctionWrapper.get().apply(a1, a2, a3, a4)[0];
                        },
                        outputDataType);
            case 5:
                return udf(
                        (a1, a2, a3, a4, a5) -> {
                            return FunctionWrapper.get().apply(a1, a2, a3, a4, a5)[0];
                        },
                        outputDataType);
            case 6:
                return udf(
                        (a1, a2, a3, a4, a5, a6) -> {
                            return FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6)[0];
                        },
                        outputDataType);
            case 7:
                return udf(
                        (a1, a2, a3, a4, a5, a6, a7) -> {
                            return FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6, a7)[0];
                        },
                        outputDataType);
            case 8:
                return udf(
                        (a1, a2, a3, a4, a5, a6, a7, a8) -> {
                            return FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6, a7, a8)[0];
                        },
                        outputDataType);
            case 9:
                return udf(
                        (a1, a2, a3, a4, a5, a6, a7, a8, a9) -> {
                            return FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6, a7, a8, a9)[0];
                        },
                        outputDataType);
            case 10:
                return udf(
                        (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10) -> {
                            return FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)[0];
                        },
                        outputDataType);
            default:
                throw new RuntimeException("UDFs with " + inputColumnNames.length + " args are not supported.");
        }
    }

    public static UserDefinedFunction createFilterFunction(String functionName, String[] inputColumnNames){
        switch(inputColumnNames.length) {
            case 0:
                return udf(
                        () -> {
                            return intToBool(FunctionWrapper.get().apply()[0]);
                        },
                        DataTypes.BooleanType);
            case 1:
                return udf(
                        (a1) -> {
                            return intToBool(FunctionWrapper.get().apply(a1)[0]);
                        },
                        DataTypes.BooleanType);
            case 2:
                return udf(
                        (a1, a2) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2)[0]);
                        },
                        DataTypes.BooleanType);
            case 3:
                return udf(
                        (a1, a2, a3) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2, a3)[0]);
                        },
                        DataTypes.BooleanType);
            case 4:
                return udf(
                        (a1, a2, a3, a4) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2, a3, a4)[0]);
                        },
                        DataTypes.BooleanType);
            case 5:
                return udf(
                        (a1, a2, a3, a4, a5) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2, a3, a4, a5)[0]);
                        },
                        DataTypes.BooleanType);
            case 6:
                return udf(
                        (a1, a2, a3, a4, a5, a6) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6)[0]);
                        },
                        DataTypes.BooleanType);
            case 7:
                return udf(
                        (a1, a2, a3, a4, a5, a6, a7) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6, a7)[0]);
                        },
                        DataTypes.BooleanType);
            case 8:
                return udf(
                        (a1, a2, a3, a4, a5, a6, a7, a8) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6, a7, a8)[0]);
                        },
                        DataTypes.BooleanType);
            case 9:
                return udf(
                        (a1, a2, a3, a4, a5, a6, a7, a8, a9) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6, a7, a8, a9)[0]);
                        },
                        DataTypes.BooleanType);
            case 10:
                return udf(
                        (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10) -> {
                            return intToBool(FunctionWrapper.get().apply(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)[0]);
                        },
                        DataTypes.BooleanType);
            default:
                throw new RuntimeException("UDFs with " + inputColumnNames.length + " args are not supported.");
        }
    }

    /**
     * Used for performance testing.
     */
    public static UserDefinedFunction createAddFunction() {
        return udf(
                (a1, a2) -> {
                    return (int)a1 + (int)a2;
                },
                DataTypes.IntegerType
        );
    }

    /**
     * Used for performance testing.
     */
    public static UserDefinedFunction createRandomFunction() {
        Random rn = new Random();
        return udf(
                (a1, a2) -> {
                    return (int)a1 + (int)a2 + rn.nextInt(10000);
                },
                DataTypes.IntegerType
        );
    }

    private static boolean intToBool(Object obj) {
        return ((Integer) obj) != 0;
    }
}

File: src/main/scala/UdfFactory.scala
//import org.apache.spark.sql.types.DataType
//import org.apache.spark.sql.functions.udf
//
//object UdfFactory {
//  def createMapFunction(inputColumnNames: Array[String], outputColumnType: String) = {
//    val outputDataType = DataType.fromDDL(outputColumnType)
//    inputColumnNames.length match {
//      case 0 =>
//        udf(() => InstanceWrapper.get.exports.getFunction().apply()(0))
//      case 1 =>
//        udf((a1: Any) => InstanceWrapper.get.apply(a1)(0))
//      case 2 =>
//        udf((a1: Any, a2: Any) => InstanceWrapper.get.apply(a1, a2)(0))
//      case 3 =>
//        udf((a1: Any, a2: Any, a3: Any) => InstanceWrapper.get.apply(a1, a2, a3)(0))
//      case 4 =>
//        udf((a1: Any, a2: Any, a3: Any, a4: Any) => InstanceWrapper.get.apply(a1, a2, a3, a4)(0))
//      case 5 =>
//        udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any) => InstanceWrapper.get.apply(a1, a2, a3, a4, a5)(0))
//      case 6 =>
//        udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any) => InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6)(0))
//      case 7 =>
//        udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any, a7: Any) => InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6, a7)(0))
//      case 8 =>
//        udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any, a7: Any, a8: Any) => InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6, a7, a8)(0))
//      case 9 =>
//        udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any, a7: Any, a8: Any, a9: Any) =>  InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6, a7, a8, a9)(0))
//      case 10 =>
//        udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any, a7: Any, a8: Any, a9: Any, a10: Any) => InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)(0))
//      case _ =>
//        throw new RuntimeException("UDFs with " + inputColumnNames.length + " args are not supported.")
//    }
//  }
//
//  def createFilterFunction(inputColumnNames: Array[String]) = inputColumnNames.length match {
//    case 0 =>
//      udf(() => intToBool(InstanceWrapper.get.apply()(0)))
//    case 1 =>
//      udf((a1: Any) => intToBool(InstanceWrapper.get.apply(a1)(0)))
//    case 2 =>
//      udf((a1: Any, a2: Any) => intToBool(InstanceWrapper.get.apply(a1, a2)(0)))
//    case 3 =>
//      udf((a1: Any, a2: Any, a3: Any) => intToBool(InstanceWrapper.get.apply(a1, a2, a3)(0)))
//    case 4 =>
//      udf((a1: Any, a2: Any, a3: Any, a4: Any) => intToBool(InstanceWrapper.get.apply(a1, a2, a3, a4)(0)))
//    case 5 =>
//      udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any) => intToBool(InstanceWrapper.get.apply(a1, a2, a3, a4, a5)(0)))
//    case 6 =>
//      udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any) => intToBool(InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6)(0)))
//    case 7 =>
//      udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any, a7: Any) => intToBool(InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6, a7)(0)))
//    case 8 =>
//      udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any, a7: Any, a8: Any) => intToBool(InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6, a7, a8)(0)))
//    case 9 =>
//      udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any, a7: Any, a8: Any, a9: Any) => intToBool(InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6, a7, a8, a9)(0)))
//    case 10 =>
//      udf((a1: Any, a2: Any, a3: Any, a4: Any, a5: Any, a6: Any, a7: Any, a8: Any, a9: Any, a10: Any) => intToBool(InstanceWrapper.get.apply(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)(0)))
//    case _ =>
//      throw new RuntimeException("UDFs with " + inputColumnNames.length + " args are not supported.")
//  }
//
//  /**
//   * Used for performance testing.
//   */
//  def createAddFunction = udf((a1: Any, a2: Any) => a1.asInstanceOf[Int] + a2.asInstanceOf[Int])
//
//
//  private def intToBool(obj: Any) = obj.asInstanceOf[Integer] ne 0
//}